orchestrator/careplancontributor/service.go

```diff
@@ -104,24 +106,63 @@ func (s Service) confirm(localFHIR fhirclient.Client, serviceRequestRef string,
 	//if err != nil {
 	//	return nil, fmt.Errorf("failed to create CarePlan: %w", err)
 	//}
-	task, err := s.createTask(*serviceRequest, patient)
+	task, err := s.createTask(*serviceRequest)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create Task: %w", err)
 	}
-	// Add Task to CarePlan
-	// Note: since we created the Task with status=accepted, we can do this immediately.
-	// Otherwise, we have to poll/wait from the UI.
-	//carePlan.Activity = append(carePlan.Activity, fhir.CarePlanActivity{
-	//	Reference: &fhir.Reference{
-	//		Type:      to.Ptr("Task"),
-	//		Reference: to.Ptr("Task/" + *task.Id), // TODO: This seems a fiddly way to reference stuff
-	//	},
-	//})
-	//// TODO: Add "If" headers to make sure we're not overwriting someone else's changes
-	//if err := s.CarePlanService.Update("CarePlan/"+*carePlan.Id, carePlan, &carePlan); err != nil {
-	//	return nil, fmt.Errorf("failed to add Task to CarePlan: %w", err)
-	//}
-	return task, err
+
+	// Create channels for polling the task, 1 channel for the result and 1 for an error
+	resultChan := make(chan *fhir.Task)
+	errorChan := make(chan error)
+
+	// Start polling in a new goroutine so that the code continues to the select below
+	go s.pollTaskStatus(*task.Id, resultChan, errorChan)
+
+	// Block until either the task is accepted or an error occurs (error after 10 attempts)
+	select {
+	case result := <-resultChan:
+		log.Info().Msgf("Task/%s accepted: %+v\n", *result.Id, result)
+		enrichedTask, err := s.handleAcceptedTask(task, serviceRequest, &patient)
+		return enrichedTask, err
+	case err := <-errorChan:
+		fmt.Printf("error while polling task %s: %v\n", *task.Id, err)
+		return nil, err
+	}
+}
+
+// pollTaskStatus polls the task status until it is accepted, an error occurs or reaches a max poll amount.
+func (s Service) pollTaskStatus(taskID string, resultChan chan<- *fhir.Task, errorChan chan<- error) {
+	ticker := time.NewTicker(2 * time.Second) // Poll every 2 seconds
+	defer ticker.Stop()
+
+	maxAttempts := 10
+	attempts := 0
+
+	for range ticker.C {
+
+		if attempts >= maxAttempts {
+			err := fmt.Errorf("maximum number of attempts reached for Task/%s", taskID)
+			log.Error().Msgf("polling Task/%s failed - error: [%s]", taskID, err)
+			errorChan <- err
+			return
+		}
+
+		attempts++
+		var task fhir.Task
+
+		if err := s.CarePlanService.Read("Task/"+taskID, &task); err != nil {
```

Comment:
This doesn't take into account the time a `Read` takes. So maybe rewrite the `for` loop with a `time.Sleep(..)` instead, and check whether the expected total time taken (now 20s) has passed? You could also do this with the ticker, but then with a `for { select (ticker.C, ctx.Done) }` (pseudo code) instead, that uses a context passed to the`pollTaskStatus` function (which can then have a time-out).

(https://golang.cafe/blog/golang-context-with-timeout-example.html)

Also; `Read` invokes an HTTP client, which has a default time-out of 30s. So this could the first iteration to time-out, right at the beginning. So ideally we specify a time-out in the `Read()` function (or actually a context which you can cancel/give a time-out). I'll add that to the library.

Reply:
Rewrote the logic so that the complete process times out after 20 seconds. I've added a todo to add a timeout to the Read when the `go-fhir-client` supports it.